engine.go

```diff
@@ -520,6 +486,100 @@ func (e *RTCEngine) configure(
 	return nil
 }
 
+func (e *RTCEngine) createSubscriberPCLocked(configuration webrtc.Configuration, isPrimary bool) error {
+	var err error
+	if e.subscriber, err = NewPCTransport(PCTransportParams{
+		Configuration:        configuration,
+		RetransmitBufferSize: e.connParams.RetransmitBufferSize,
+	}); err != nil {
+		return err
+	}
+	e.subscriber.SetLogger(e.log)
+
+	e.subscriber.OnRemoteDescriptionSettled(e.createSubscriberPCAnswerAndSend)
+
+	e.subscriber.pc.OnICECandidate(func(candidate *webrtc.ICECandidate) {
+		if candidate == nil {
+			// done
+			return
+		}
+		init := candidate.ToJSON()
+		e.log.Debugw(
+			"local ICE candidate",
+			"target", livekit.SignalTarget_SUBSCRIBER,
+			"candidate", init.Candidate,
+		)
+		if err := e.signalTransport.SendMessage(
+			e.signalling.SignalICECandidate(
+				protosignalling.ToProtoTrickle(init, livekit.SignalTarget_SUBSCRIBER, false),
+			),
+		); err != nil {
+			e.log.Errorw("could not send ICE candidates for subscriber", err)
+		}
+	})
+
+	e.subscriber.pc.OnICEConnectionStateChange(func(state webrtc.ICEConnectionState) {
+		e.handleICEConnectionStateChange(e.subscriber, livekit.SignalTarget_SUBSCRIBER, isPrimary, state)
+	})
+
+	e.subscriber.pc.OnTrack(func(remote *webrtc.TrackRemote, receiver *webrtc.RTPReceiver) {
+		e.engineHandler.OnMediaTrack(remote, receiver)
+	})
+
+	e.subscriber.pc.OnDataChannel(func(c *webrtc.DataChannel) {
+		e.dclock.Lock()
+		defer e.dclock.Unlock()
+		if c.Label() == reliableDataChannelName {
+			e.reliableDCSub = c
+		} else if c.Label() == lossyDataChannelName {
+			e.lossyDCSub = c
+		} else {
+			return
+		}
+		c.OnMessage(e.handleDataPacket)
+	})
+
+	return nil
+}
+
+func (e *RTCEngine) handleICEConnectionStateChange(
+	transport *PCTransport,
+	signalTarget livekit.SignalTarget,
+	isPrimary bool,
+	state webrtc.ICEConnectionState,
+) {
+	switch state {
+	case webrtc.ICEConnectionStateConnected:
+		var fields []interface{}
+		if pair, err := transport.GetSelectedCandidatePair(); err == nil {
+			fields = append(fields, "transport", signalTarget, "iceCandidatePair", pair)
+		}
+		e.log.Debugw("ICE connected", fields...)
+	case webrtc.ICEConnectionStateDisconnected:
+		e.log.Debugw("ICE disconnected", "transport", signalTarget)
+	case webrtc.ICEConnectionStateFailed:
+		e.log.Debugw("ICE failed", "transport", signalTarget)
+		if isPrimary {
+			e.handleDisconnect(false)
```

Comment:
should we handle non-primary disconnect as well?

Reply:
Sorry, missed this before merging @cnderrauber . Yeah, good point. No reason to not handle it. I just repeated what was there. Will change it in next commit to handle it for both.